File: Random_walk.py
import random
import time

def random_walk(n):
    '''
    return the distance abs(x) + abs(y) after n random walk on a 2-d map
    '''
    x, y = 0, 0
    for i in range(n):
        dx, dy = random.choice([(0,1), (0,-1), (1,0), (-1,0)])
        x += dx
        y += dy
    distance = abs(x) + abs(y)
    return distance

if __name__ == '__main__':
    '''
    if the distance is less or equal than 4, the man needn't take transport home 
    '''
    walk_length = 30
    sample_size = 100000
    for step in range(1, walk_length):
        tick = time.perf_counter()
        no_of_no_transport = 0
        for i in range(sample_size):
            distance = random_walk(step)
            if distance <= 4:
                no_of_no_transport += 1
        percentage = float(no_of_no_transport / sample_size) * 100
        print(f'walk_lenth is {step}, the rate of no_transportation is {percentage}')
        tock = time.perf_counter()
        print(f'walk_length: {step} - simple size: {sample_size} - time consume: {1e3*(tock-tick):.6f} ms')
    
        

File: random_walk.cpp
#include <iostream>
#include <random>
#include <chrono>
using namespace std;
using namespace std::chrono;

mt19937 gen{random_device{}()};
uniform_int_distribution<int> dist{1,4};

int random_walk(const int& n){
    int x = 0;
    int y = 0;
    for (auto i = 0; i<n; i++){
        int direction = dist(gen);
        switch (direction)
        {
            case 1:
                x += 1;
                break;
            case 2:
                x -= 1;
                break;
            case 3:
                y += 1;
                break;
            case 4:
                y -= 1;
                break;
        }
    }
    int distance = abs(x) + abs(y);
    return distance; 
}

int main() {
    int steps = 30;
    int sample_size = 100000;
    for(auto step=1; step<steps; step++){
        auto start = high_resolution_clock::now();
        int no_of_stransport = 0;
        for(auto test=0; test<sample_size; test++){
            int distance = random_walk(step);
            if(distance <= 4)
                no_of_stransport += 1;
        }
        float percentage = float(no_of_stransport)/float(sample_size) * 100;
        cout<<"walk_lenth is "<<step<<", the rate of no_transportation is "<<percentage<<endl;
        auto stop = high_resolution_clock::now();
        auto duration = duration_cast<milliseconds>(stop - start).count();
        cout<<"time consume: "<<duration<<" ms"<<endl;
    }
}